config/base.go

```diff
@@ -0,0 +1,81 @@
+// Copyright © 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// BaseConfig is a generic interface for TOML configs in mixer
+type BaseConfig interface {
+	Save() error
```

Comment:
Any reason not to add `SetProperty` and `GetProperty` to this interface?

Reply:
Also `Load`? (instead of `LoadConfig`, which would be a stutter)

Reply:
Save is the only item in the interface because it is the only shared method needed for the feature. Moving into a more generic definition, `Load` would be a nice inclusion.

As for `SetProperty` and `GetProperty`, they have a single implementation for all sub-classes, unlike `Load`and `Save` that are subclass specific. Ideally they would be methods of the `BaseConfig` class and the subclasses would just inherit them (instead of overloading them), but since there is no such a thing in go, we keep them out of the interface.

Reply:
example: https://github.com/matthewrsj/encap

Reply:
So, If I get this right, your suggestion is to have `ConfigInterface` which define the methods, an empty `Config` structure that allows me to implement `SetProperty` and `GetProperty` and the other classes that will include `Config` as a property. Is it right?

Reply:
Right but I wouldn't call it `ConfigInterface`, probably just `Config`
```go
type Config interface {
    Save() error
    Load(string) error
    SetProperty(string, string) error
    GetProperty(string) (string, error)
}

type Base struct {
    // whatever
}

func (b *Base) Save() error {
}

func (b *Base) Load(s string) error {
}

func (b *Base) SetProperty(k, v string) error {
}

func (b *Base) GetProperty(k string) (string, error) {
}

type MixConfig struct {
    Base
}

// we need a different for GetProperty for MixConfig for whatever reason
func (m *MixConfig) GetProperty(k string) (string, error) {
}
```
(I might have gotten some of the function sigs wrong)
This allows you to call `m.SetProperty` since it is defined on the embedded base struct. Basically just a different syntax for class inheritance.

Reply:
And that `Base` struct wouldn't necessarily have to be empty. Ideally it should have all the common members between all subclasses, if there is commonality.

Reply:
Well, to be honest I'm not convinced with this solution. It seems to make it much more complicated than it needs to be. You get one extra layer of abstraction which makes the code harder to follow and get little benefit out of it.

My personal opinion is that it is better to call `SetProperty(m)` instead of having all this code to call `m.SetProperty()`.

Reply:
Ok,  I understand the reasoning for keeping `SetPropery()` and `GetProperty() `separate and having the first parameter be a Config interface, that works fine, it means we can pass _any_ config into them. The `save` & `load` should be both declared in the interface though since they may have diff implementations, and the interface should be renamed to just `type Configuration interface`.  It sounds ok to name things "base" until you expand your implementation and support to where that name becomes ambiguous/wrong. You want to maintain an "is-a" relationship between objects, but not through their naming IMHO.